segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java

```diff
@@ -829,6 +863,37 @@ private CompletableFuture<EpochInfo> readEpochInfo(String chunk, ChunkedSegmentS
 
     }
 
+    /**
+     * Read the container epoch saved in "container_<containerid>_epoch file
+     * saved on storage that is created while generating the backup.
+     * Container startup in case of recovery mode uses this saved epoch information
+     * @return epoch read from storage.
+     */
+    private CompletableFuture<Long> readContainerEpoch() {
+        log.info(" {}: Reading container epoch from storage", this.traceObjectId);
+        val containerEpochFileName = NameUtils.getContainerEpochFileName(this.getId());
+        UtilsWrapper wrapper = new UtilsWrapper((ChunkedSegmentStorage) this.storage,BUFFER_SIZE, this.config.getMetadataStoreInitTimeout() );
+        ByteBufferOutputStream outputStream = new ByteBufferOutputStream();
+        return wrapper.copyFromChunk(containerEpochFileName, outputStream)
+                .thenComposeAsync( v -> {
+                    EpochInfo containerEpoch;
+                    try {
+                        containerEpoch = EPOCH_INFO_SERIALIZER.deserialize(outputStream.getData());
+                    }catch(Exception e) {
+                        throw new RuntimeException(e);
+                    }
+                    log.info("{}: Read container epoch {} from storage", this.traceObjectId, containerEpoch.getEpoch());
+                    return CompletableFuture.completedFuture(containerEpoch.getEpoch() + 1);
```

Comment:
Not sure I understand whether adding 1 is sufficient guarantee that epoch will be unique. Especially with multiple containers.

Reply:
this just adds 1 to the container epoch in the "recover" path just like in the initialize(non-recover) path.